dockerfile_parse/parser.py

```diff
@@ -301,6 +321,10 @@ def parent_images(self):
                 continue
             image, _ = image_from(instr['value'])
             if image is not None:
+                # replace any ARG keys with values
```

Comment:
I don't think that this is proper solution. ARGs may be used on many places, you covered only one case.
Also parser should return exact content of dockerfile if not specified otherwise.

I'd go with a separate method which replaces ARGs within dockerfile and returns new parser object with replaced occurrences of ARGS.

WDYT?

Reply:
Respectively for ENVs there is:
```
self.env_replace = env_replace
```
in init. We can follow the same logic as was used for envs with args as they are basically the same only used in different context.

Reply:
> I don't think that this is proper solution. ARGs may be used on many places, you covered only one case.

To be fair, it's not like the current implementation supports env replacement anywhere other than the `envs` and `labels` properties. If we want to support all instructions, it might be best to use a different approach (such as modifying the `_instruction_getter()` method to support every type of instruction or incorporating the `context_structure` property).

I'm not sure if the intention of this PR was to support env/arg replacement in FROM instructions but support for ARG handling will nevertheless need some more work.

Reply:
Agree that this is a quick n' dirty replacement in FROM mostly to address the specific issue in #74 that ARG values are not replaced in FROM. If you would rather merge in a PR that supports ARG completey I could work on that.

Reply:
I'd like to have ARGs and ENVs aligned handling. No need for extra functionality which is not supported by ENVs already (Except that FROM replacement).

Reply:
> I don't think that this is proper solution. ARGs may be used on many places, you covered only one case.
> Also parser should return exact content of dockerfile if not specified otherwise.
> 
> I'd go with a separate method which replaces ARGs within dockerfile and returns new parser object with replaced occurrences of ARGS.
> 
> WDYT?

What I am noticing is that env replacement happens just for ENV or LABEL instructions and not for RUN instructions nor anywhere in the dockerfile content itself. So I implemented the ARG getter the same way as the ENV getter which uses `_instruction_getter()`. The replacement in the FROM line is to address the specific issue reported in #74. I can do a separate PR for global content replacement which will remove this change. Or maybe I'm missing something in the implementation...

Reply:
> Respectively for ENVs there is:
> 
> ```
> self.env_replace = env_replace
> ```
> 
> in init. We can follow the same logic as was used for envs with args as they are basically the same only used in different context.

Taking a step back here: What does `env_replace` do? It doesn't say in the docstring.

Reply:
`env_replace` is a flag which decides if ENV variables will be  expanded or not. However, as you rightly pointed out, this variable expansion currently only supports ENV and LABEL instructions.

I think what we want is for ARGs to be treated the same as ENVs, as in support ARG expansion wherever ENV expansion is currently supported etc. I'm afraid you will find a lot of the logic required for that is not yet in place though.

If you would rather only support ARG expansion in FROM instructions then that is fine too, as long as it respects the `env_replace` flag (or a new one specifically for this) and is properly documented.

Reply:
@MartinBasti @chmeliik Sorry for the late reply.
The trouble with maintaining the same logic for ARGs and ENVs with variable expansion is that ARGs are not used the same way as ENVs. If I apply the same ENVs expansion to ARGs in the Dockerfile examples from https://docs.docker.com/engine/reference/builder/#arg
they end up looking like this:
```
# from example: https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact
ARG VERSION=latest
FROM busybox:$VERSION
ARG VERSION
RUN echo $VERSION > image_version
 
if env_replace is True:
-> {'VERSION': 'latest'}

if env_replace is False:
-> {'VERSION': 'latest', 'VERSION': ''} (Is this expected behaviour?)
```
It also seems to me you don't want to fix #74 (ARG replacement in the FROM instruction)?

Reply:
Sorry for late reply. I'm not exactly sure what you mean. What is difference between ARGs and ENVs except that ARGs are available only during buildtime

Reply:
Unlike ENVs, ARGs don't have to have a value for the variable in the Dockerfile (see `ARG VERSION` in the example). I am trying to understand what the expected behaviour is in the case where it doesn't have a variable and `env_replace` is True or False.

Reply:
ENVs also don't necessarily have a value defined in the Dockerfile -- it might come from the parent image's config. They feel quite similar to me.